File: app/src/main/java/pl/revanmj/smspasswordnotifier/data/WhitelistAdapter.java
package pl.revanmj.smspasswordnotifier.data;

import android.graphics.Color;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.util.SparseIntArray;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import pl.revanmj.smspasswordnotifier.R;

/**
 * Created by revanmj on 15.03.2017.
 */

public class WhitelistAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
    private static final int EMPTY_VIEW = 10;
    private List<WhitelistItem> mWhitelist;
    private ClickListener mClickListener;
    private SparseIntArray mSelectedItems = new SparseIntArray();

    public WhitelistAdapter(ClickListener listener) {
        mClickListener = listener;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {

        if (viewType == EMPTY_VIEW) {
            View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.empty_view, parent, false);
            return new EmptyViewHolder(v);
        }

        View itemView = LayoutInflater.
                from(parent.getContext()).
                inflate(R.layout.edit_whitelist_item, parent, false);

        return new WhitelistItemHolder(itemView);
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        if (holder instanceof WhitelistItemHolder) {
            WhitelistItem item = getItem(position);
            ((WhitelistItemHolder) holder).bind(
                    item.getSenderId(),
                    item.getName());
            holder.itemView.setOnClickListener(view -> {
                if (mClickListener != null) {
                    mClickListener.onItemClicked(position);
                }
            });
            holder.itemView.setOnLongClickListener(view -> {
                if (mClickListener != null) {
                    return mClickListener.onItemLongClicked(position);
                }
                return false;
            });
            if (isSelected(position))
                ((WhitelistItemHolder)holder).hightlightItem();
            else
                ((WhitelistItemHolder)holder).unhighlightItem();
        }
    }

    @Override
    public int getItemCount() {
        if (mWhitelist != null && mWhitelist.size() > 0)
            return mWhitelist.size();
        else
            return 1;
    }

    @Override
    public int getItemViewType(int position) {
        if (mWhitelist == null || mWhitelist.size() == 0) {
            return EMPTY_VIEW;
        }
        return super.getItemViewType(position);
    }

    public void swapData(List<WhitelistItem> whitelist) {
        mWhitelist = whitelist;
        notifyDataSetChanged();
    }

    private WhitelistItem getItem(int pos) {
        return mWhitelist.get(pos);
    }

    public boolean isSelected(int position) {
        return mSelectedItems.get(position, -1) >= 0;
    }

    public void toggleSelection(int position) {
        if (mSelectedItems.get(position, -1) >= 0) {
            mSelectedItems.delete(position);
        } else {
            mSelectedItems.put(position, getItem(position).getSenderId());
        }
        notifyItemChanged(position);
    }

    public int getSelectedItemCount() {
        return mSelectedItems.size();
    }

    public List<WhitelistItem> getSelectedItems() {
        List<WhitelistItem> result = new ArrayList<>();
        for (int i = 0; i < mSelectedItems.size(); i++) {
            result.add(mWhitelist.get(mSelectedItems.keyAt(i)));
        }
        return result;
    }

    public void clearSelection() {
        for (int i = 0; i < mSelectedItems.size(); ++i) {
            notifyItemChanged(mSelectedItems.keyAt(i));
        }
        mSelectedItems.clear();
    }

    public class WhitelistItemHolder extends RecyclerView.ViewHolder {
        private TextView sender;
        private int _id;
        private LinearLayout linearLayout;

        WhitelistItemHolder(View itemView) {
            super(itemView);

            sender = itemView.findViewById(R.id.sender);
            linearLayout = itemView.findViewById(R.id.linear_layout);
        }

        void bind(int id, String sender) {
            this._id = id;
            this.sender.setText(sender);
        }

        public int getId() {
            return _id;
        }

        void hightlightItem() {
            linearLayout.setBackgroundResource(R.color.selected);
        }

        void unhighlightItem() {
            linearLayout.setBackgroundColor(Color.TRANSPARENT);
        }
    }

    public class EmptyViewHolder extends RecyclerView.ViewHolder {
        EmptyViewHolder(View itemView) {
            super(itemView);
        }
    }

    public interface ClickListener {
        void onItemClicked(int position);
        boolean onItemLongClicked(int position);
    }
}

File: app/src/main/java/pl/revanmj/smspasswordnotifier/data/WhitelistAdapter.kt
package pl.revanmj.smspasswordnotifier.data

import android.graphics.Color
import android.util.SparseIntArray
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView

import androidx.recyclerview.widget.RecyclerView

import java.util.ArrayList

import pl.revanmj.smspasswordnotifier.R

/**
 * Created by revanmj on 15.03.2017.
 */

class WhitelistAdapter(private val mClickListener: ClickListener?) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    companion object {
        private const val EMPTY_VIEW = 10
    }

    private var mWhitelist: List<WhitelistItem>? = null
    private val mSelectedItems = SparseIntArray()

    val selectedItemCount: Int
        get() = mSelectedItems.size()

    val selectedItems: List<WhitelistItem>
        get() {
            val result = ArrayList<WhitelistItem>()
            for (i in 0 until mSelectedItems.size()) {
                result.add(mWhitelist!![mSelectedItems.keyAt(i)])
            }
            return result
        }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        if (viewType == EMPTY_VIEW) {
            val v = LayoutInflater.from(parent.context).inflate(R.layout.empty_view, parent, false)
            return EmptyViewHolder(v)
        }

        val itemView = LayoutInflater.from(parent.context).inflate(R.layout.edit_whitelist_item, parent, false)
        return WhitelistItemHolder(itemView)
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (holder is WhitelistItemHolder) {
            val item = getItem(position)
            holder.bind(
                    item.senderId,
                    item.name)
            holder.itemView.setOnClickListener {
                mClickListener?.onItemClicked(position)
            }
            holder.itemView.setOnLongClickListener {
                if (mClickListener != null)
                    return@setOnLongClickListener mClickListener.onItemLongClicked(position)
                false
            }
            if (isSelected(position))
                holder.highlightItem()
            else
                holder.unhighlightItem()
        }
    }

    override fun getItemCount(): Int {
        return if (mWhitelist != null && mWhitelist!!.isNotEmpty())
            mWhitelist!!.size
        else
            1
    }

    override fun getItemViewType(position: Int): Int {
        return if (mWhitelist == null || mWhitelist!!.isEmpty()) {
            EMPTY_VIEW
        } else super.getItemViewType(position)
    }

    fun swapData(whitelist: List<WhitelistItem>) {
        mWhitelist = whitelist
        notifyDataSetChanged()
    }

    private fun getItem(pos: Int): WhitelistItem {
        return mWhitelist!![pos]
    }

    private fun isSelected(position: Int): Boolean {
        return mSelectedItems.get(position, -1) >= 0
    }

    fun toggleSelection(position: Int) {
        if (mSelectedItems.get(position, -1) >= 0) {
            mSelectedItems.delete(position)
        } else {
            mSelectedItems.put(position, getItem(position).senderId)
        }
        notifyItemChanged(position)
    }

    fun clearSelection() {
        selectedItems.forEachIndexed {
            index, _ -> notifyItemChanged(mSelectedItems.keyAt(index))
        }
        mSelectedItems.clear()
    }

    inner class WhitelistItemHolder internal constructor(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val sender: TextView = itemView.findViewById(R.id.sender)
        private val linearLayout: LinearLayout = itemView.findViewById(R.id.linear_layout)
        var id: Int = 0
            private set

        internal fun bind(id: Int, sender: String?) {
            this.id = id
            this.sender.text = sender
        }

        internal fun highlightItem() {
            linearLayout.setBackgroundResource(R.color.selected)
        }

        internal fun unhighlightItem() {
            linearLayout.setBackgroundColor(Color.TRANSPARENT)
        }
    }

    inner class EmptyViewHolder internal constructor(itemView: View) : RecyclerView.ViewHolder(itemView)

    interface ClickListener {
        fun onItemClicked(position: Int)
        fun onItemLongClicked(position: Int): Boolean
    }
}
